charms/tensorboard-controller/src/charm.py

```diff
@@ -1,184 +1,302 @@
+        layer_config = {
+            "summary": "tensorboard-controller layer",
+            "description": "Pebble config layer for tensorboard-controller",
+            "services": {
+                self._container_name: {
+                    "override": "replace",
+                    "summary": "Entrypoint of tensorboard-controller image",
+                    "command": self._exec_command,
+                    "startup": "enabled",
+                    "environment": self.service_environment,
+                    "on-check-failure": {"tensorboard-controller-up": "restart"},
+                }
             },
-            {
-                "kubernetesResources": {
-                    "customResourceDefinitions": [
-                        {"name": crd["metadata"]["name"], "spec": crd["spec"]}
-                        for crd in yaml.safe_load_all(Path("files/crds.yaml").read_text())
-                    ],
-                },
+            "checks": {
+                "tensorboard-controller-up": {
+                    "override": "replace",
+                    "period": "30s",
+                    "timeout": "20s",
+                    "threshold": 4,
+                    "http": {"url": f"http://localhost:{PROBE_PORT}{PROBE_PATH}"},
+                }
             },
-        )
+        }
 
-        self.model.unit.status = ActiveStatus()
+        return Layer(layer_config)
+
+    def _check_and_report_k8s_conflict(self, error) -> bool:
+        """Return True if error status code is 409 (conflict), False otherwise."""
+        if error.status.code == 409:
+            self.logger.warning(f"Encountered a conflict: {error}")
+            return True
+        return False
+
+    def _apply_k8s_resources(self, force_conflicts: bool = False) -> None:
+        """Apply K8s resources.
+
+        Args:
+            force_conflicts (bool): *(optional)* Will "force" apply requests causing conflicting
+                                    fields to change ownership to the field manager used in this
+                                    charm.
+                                    NOTE: This will only be used if initial regular apply() fails.
+        """
+        self.unit.status = MaintenanceStatus("Creating K8s resources")
+        try:
+            self.k8s_resource_handler.apply()
+        except ApiError as error:
+            if self._check_and_report_k8s_conflict(error) and force_conflicts:
+                # conflict detected when applying K8s resources
+                # re-apply K8s resources with forced conflict resolution
+                self.unit.status = MaintenanceStatus("Force applying K8s resources")
+                self.logger.warning("Apply K8s resources with forced changes against conflicts")
+                self.k8s_resource_handler.apply(force=force_conflicts)
+            else:
+                raise GenericCharmRuntimeError("K8s resources creation failed") from error
```

Comment:
(I realise I think you inherritted this from somewhere else)

I wonder if we can encapsulate and reuse this better.  Essentially we do the same thing twice.  Maybe this can be in a descriptive helper (`apply_and_force_if_needed`?  that sounds dumb... :D ) that means we only have the code once, and a first-time reader can understand it better?

Reply:
For starters, I hate that I even have to copy paste those handlers. I believe it would make more sense for us to discuss how we can implement and maintain them in a single place, rather than have different versions of the exact same thing across our repos.

Reply:
yep agreed with both of you.  If we have a common pattern, we should centralise it.  And what @dnplas proposes makes sense to me, although I'm open to any ideas that get it done